Sync schemas and add support for viewing global schema

shahzebsiddiqui opened the pull request:
This PR is adding the following support
- Sync global and script schema 
- Option ``buildtest show schema --global`` to view the global schema. I figure this would be special schema since it doesn't have version and up a directory as pose to script schema
- Set --name to use choices field that allows argparse to handle to input choice which makes sense to me.
- Lastly we increase coverage to 100% for buildtest/menu/show.py

shahzebsiddiqui committed:
Sync script and global schema from https://github.com/buildtesters/schemas
Set buildtest show schema --name to use choices field to restrict search
to supported schemas.
add option buildtest show schema --global to view global schema
Add coverage to 100% for file buildtest/menu/show.py

vsoch commented on buildtest/menu/__init__.py:
```diff
@@ -183,6 +184,17 @@ def show_menu(self):
+            help="show schema by name (e.g., script)",
+            choices=supported_schemas,
+        )
+        # can't use attribute args.global so setting dest=main for argparse
```
Maybe just do args._global instead? It would still be intuitive , remove the need for this comment, and not trigger the error.

vsoch commented on buildtest/menu/show.py:
```diff
@@ -1,28 +1,38 @@
+       :result: output of json schema on console
+    """
+
+    # implements ``buildtest show schema --global``. Since we can't use args.global we set this to main
```
Same here, args._global is clear.

shahzebsiddiqui replied:
just to be clear you want to change ``dest=name`` to ``dest=_global`` to resemble the name of the option. I couldn't understand is ``args.global`` a reserved attribute in argparse

vsoch commented on buildtest/menu/show.py:
```diff
@@ -1,28 +1,38 @@
+        global_schema = os.path.join(here, "global.schema.json")
+        # check if file exists
+        assert global_schema
+        recipe = load_schema(global_schema)
```
This should be refactored to get some config schema depending on the args provided, and then one line at the end to print. If you see a pattern that has redundant code and an early return it’s usually an indication that you can implement it more succinctly.

shahzebsiddiqui replied:
I realized that I couldn't use ``get_schemas_available`` https://github.com/shahzebsiddiqui/buildtest-1/blob/schema_updates/buildtest/buildsystem/schemas/utils.py#L104 since it's relied on a specific directory structure like ``script/script-v0.0.1.schema.json``.

My thought was to introduce a separate option ``--global`` instead of using ``--name global``. I could separate out the functionality into separate methods one for name schema (``compiler``, ``script``) and the other for global schema.

Unrelated to this but i see that settings schema is in a different path https://github.com/buildtesters/buildtest/blob/devel/buildtest/settings/settings.schema.json  apart from other schemas and it was intentional.  This means the implementation depends on the directory structure and where all the schemas need to be placed.

I had some other ideas in mind but didn't introduce them yet for example
- Display settings schema https://github.com/buildtesters/buildtest/blob/devel/buildtest/settings/settings.schema.json maybe it could be ``buildtest show schema --settings``. I wasn't sure if it made sense to group this feature in ``buildtest config`` with some option like ``buildtest config schema``
- Add support to fetch development schemas such as https://buildtesters.github.io/schemas/script/script-v0.0.1.schema.json and others.

vsoch replied:
You are misunderstanding. I'm not talking about how it's loaded, but your instantiation of the client here.

vsoch commented on buildtest/menu/show.py:
```diff
@@ -1,28 +1,38 @@
+        global_schema = os.path.join(here, "global.schema.json")
+        # check if file exists
+        assert global_schema
+        recipe = load_schema(global_schema)
```
You are doing:
```
if this evaluates to true:
    set file to be this
    load
    print
    return

set file to be this
load
print
```

and what you should do for less redundant code is

```
determine file based on input args, set file to be this
load
print
```
Hopefully that pseudo code helps :)

shahzebsiddiqui replied:
Yea i see your point thanks for the hint, i will make some adjustment and get back to  you.

vsoch replied:
Great! Design details like this are important, you will over time become a better programmer. :)

shahzebsiddiqui committed:
change dest attribute to _global  for option buildtest show schema
--global
Refactor method 'show_schema_layout' by removing duplication of code
since --global and --name are doing the same operation but on different
schema files. Now we check the file name in advance in if/elif block and
then load the schema and print the content.
